src/components/DeeplinkWrapper/index.website.js

```diff
@@ -1,41 +1,90 @@
     children: PropTypes.node.isRequired,
+    /** User authentication status */
+    isAuthenticated: PropTypes.bool.isRequired,
 };
 
-class DeeplinkWrapper extends PureComponent {
-    componentDidMount() {
-        if (!this.isMacOSWeb() || CONFIG.ENVIRONMENT === CONST.ENVIRONMENT.DEV) {
-            return;
+function isMacOSWeb() {
+    return !Browser.isMobile() && typeof navigator === 'object' && typeof navigator.userAgent === 'string' && /Mac/i.test(navigator.userAgent) && !/Electron/i.test(navigator.userAgent);
+}
+
+function promptToOpenInDesktopApp() {
+    // If the current url path is /transition..., meaning it was opened from oldDot, during this transition period:
+    // 1. The user session may not exist, because sign-in has not been completed yet.
+    // 2. There may be non-idempotent operations (e.g. create a new workspace), which obviously should not be executed again in the desktop app.
+    // So we need to wait until after sign-in and navigation are complete before starting the deeplink redirect.
+    if (Str.startsWith(window.location.pathname, Str.normalizeUrl(ROUTES.TRANSITION_BETWEEN_APPS))) {
+        App.beginDeepLinkRedirectAfterTransition();
+    } else {
+        App.beginDeepLinkRedirect();
+    }
+}
+function DeeplinkWrapper({children, isAuthenticated}) {
+    const [currentScreen, setCurrentScreen] = useState();
+    const [hasShownPrompt, setHasShownPrompt] = useState(false);
+    const removeListener = useRef();
+    // CONFIG.ENVIRONMENT = CONST.ENVIRONMENT.STAGING;
+
+    useEffect(() => {
+        // If we've shown the prompt and still have a listener registered,
+        // remove the listener and reset its ref to undefined
+        if (hasShownPrompt && removeListener.current !== undefined) {
+            removeListener.current();
+            removeListener.current = undefined;
         }
 
-        // If the current url path is /transition..., meaning it was opened from oldDot, during this transition period:
-        // 1. The user session may not exist, because sign-in has not been completed yet.
-        // 2. There may be non-idempotent operations (e.g. create a new workspace), which obviously should not be executed again in the desktop app.
-        // So we need to wait until after sign-in and navigation are complete before starting the deeplink redirect.
-        if (Str.startsWith(window.location.pathname, Str.normalizeUrl(ROUTES.TRANSITION_BETWEEN_APPS))) {
-            App.beginDeepLinkRedirectAfterTransition();
+        if (isAuthenticated === false) {
+            setHasShownPrompt(false);
+            Navigation.isNavigationReady().then(() => {
+                // get initial route
+                const initialRoute = navigationRef.current.getCurrentRoute();
+                setCurrentScreen(initialRoute.name);
+
+                removeListener.current = navigationRef.current.addListener('state', () => {
+                    setCurrentScreen(Navigation.getActiveRouteName());
```

Comment:
@lindboe let me know if this is more in line with what you were thinking for that TODO. It's not grabbing the route in quite the same way but in testing it resulted in the same route name being set.

Reply:
I do think we should be getting it from the event data. I would worry about timing issues with using the listener, but using a hook instead of the event data the listener provides (e.g., react-navigation could be refactored and the event could fire before state used by the hook actually updates)

Reply:
It's not a huge deal to do it like this, but it's surprising and not how the listener is intended to be used, so a bit of a concern to me

Reply:
I guess maybe I misunderstood the point of the comment, did we want to move all of the functionality of adding the listeners and using the events provided by it to the `Navigation` lib or just the way in which we're accessing the shape of the state object to ensure consistency if it changes?

Reply:
> did we want to move all of the functionality of adding the listeners and using the events provided by it to the Navigation lib or just the way in which we're accessing the shape of the state object to ensure consistency if it changes?

I think the latter part is the most essential, but a function to add a listener would be fine (but more complex).

Making more complicated property access (e.g., going 3+ levels and including an array indexing operation) into a named utility function (like `getCurrentRouteFromEvent`) is just the principle I try to follow for two reasons: 1. Naming what you're trying to do in case it's confusing, and 2. Standardizing approach to reduce bugs (in case anyone else needs to do this)